### DISCUSSION TITLE
Svelte(Kit), TypeScript and incorrectly ArgTypes

### DISCUSSION THREAD

**Discussion by pepa-linha** [storybookjs/storybook#21826] [Help] [answered]
Hi, is there currently a solution (perhaps a workaround) how to generate documentation to use TypeScript types? At the moment, if I have a <script lang="ts">...</script> component in Svelte, the props description is not even generated.

I read about `react-docgen` that it should be able to TypeScript, but that didn't work for me either.

---
### Fixed

It was fixed in https://github.com/storybookjs/storybook/pull/24096 by https://github.com/j3rem1e, thanks!

**Comment by JReinhold** [comment, marked as answer]
Unfortunately not. Storybook uses [`sveltedoc-parser`](https://github.com/alexprey/sveltedoc-parser) under the hood to infer argTypes from Svelte component props, and that library only supports JSDocs, not TypeScript types. There's no tool in the Svelte ecosystem that supports getting prop types from TypeScript definitions.

So you have to add JSDocs to your props if you want automatic inference (you can still keep the TS types though).

Here's an example component with JSDocs for props: https://github.com/alexprey/sveltedoc-parser/blob/dev/examples/Button.svelte

**Reply by pepa-linha** [reply]
Thanks, JSDoc would be quite a sufficient solution, but if the script tag is written as `<script lang="ts">`, then JSDoc doesn't work either. I wouldn't want to remove support for TypeScript because of this.

**Reply by JReinhold** [reply]
Oh yeah you're right, I didn't actually know that. That's unfortunate.

**Reply by TrungRueta** [reply]
@JReinhold @pepa-linha 
I just want to confirm something following your suggestion. 

First, your suggestion is correct! Using `jsdoc` really helps StoryBook auto-detect the args type.  But it not only working with svelte js `<script />` tag. it also work with `<script lang="ts" />` tag.

This is my component write in SvelteKit and use JsDoc to define types:

```svelte
<script lang="ts">
	/** @type {string|undefined} */
	export let imgSrc: string | undefined = undefined;

	/** @type {string} */
	export let title: string;

	/** @type {string|undefined} */
	export let href: string | undefined = undefined;

	/** @type {function} */
	export let onDropdown: (() => unknown) | undefined = undefined;
...
```

And this is how StoryBook can read:

<img width="1054" alt="image" src="https://github.com/storybookjs/storybook/assets/12226007/ad88adcd-7f8d-49d0-992e-c23ba29ab507">

Hope can clarify issue and help everybody got same case!

**Reply by JReinhold** [reply]
Thanks for pointing this out @TrungRueta ! This was acutally a fix that landed recently in Storybook 7.5.0

https://github.com/storybookjs/storybook/releases/tag/v7.5.0

**Comment by brittneypostma** [comment]
Using SvelteKit package seems to strip the jsdoc types and comments from the components so this isn't working currently for me. Is there no way to override the types seen through a Storybook prop?

**Reply by JReinhold** [reply]
I'm not sure exactly what you're experiencing, but here's a StackBlitz where descriptions and types are inferred only from the JSDoc types, even with `lang="ts"`. I've removed any other hints that Storybook uses (default values, argTypes).

If you remove the JSDocs and keep the TS types (and reload the Storybook view) you'll see that the Controls are no longer correct, so this would indicate that it uses the JSDocs for arg types inference.

https://stackblitz.com/~/edit/github-rlf4nm?file=src%2Fstories%2FButton.svelte
